packages/docregistry/src/default.ts

```diff
@@ -508,6 +511,18 @@ export class DocumentWidget<
     /* no-op */
   }
 
+  /**
+   * Handle a title change.
+   */
+  private _onTitleChanged(_sender: Title<this>) {
+    const validNameExp = /[\/\\:]/;
+    const name = this.title.label;
+    const filename = this.context.path.split('/').pop()!;
+    if (name !== filename && name.length > 0 && !validNameExp.test(name)) {
+      void this.context.rename(name);
+    }
```

Comment:
We should probably reset the title back to the filename if we don't end up calling the rename, something like:
```suggestion
    if (name === filename) {
      return;
    }
    if (name.length > 0 && !validNameExp.test(name)) {
      void this.context.rename(name);
    } else {
      this.title.label = filename
    }
```
We probably also want to reset if the rename does not succeed (i.e., wait for the rename, and in case of error reset the title).
```